lib/APPTurtle/src/Transports/CustomRosTransportTcp.cpp

```diff
@@ -0,0 +1,337 @@
+        {
+            LOG_ERROR("Write prefix error (size=%zu, sent=%zu)", size, sent);
+            *errorCode = 1;
+        }
+        else if ((sent = m_tcpClient.write(buffer, size)) != size)
+        {
+            LOG_ERROR("Write data error (size=%zu, sent=%zu)", size, sent);
+            *errorCode = 1;
+        }
+        else
+        {
+            *errorCode = 0;
+        }
+    }
+
+    return sent;
+}
+
+size_t CustomRosTransportTcp::read(uint8_t* buffer, size_t size, int timeout, uint8_t* errorCode)
+{
+    size_t readBytes = 0U;
+
+    if ((nullptr == buffer) || (0U == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        bool loop = true;
+
+        /**
+         * Run receive state machine. 
+         * The read function relationships are:
+         * 
+         * read() -> [state]->read() -> readInternal() -> WifiClient
+         */
+        while (loop)
+        {
+            loop = (this->*m_readFunction[m_inputState])(timeout, errorCode);
+        }
+
+        if (InputState::FINISH == m_inputState)
+        {
+            if (readBytes > size)
+            {
+                /* internal error, request buffer would be overrun */
+                close();
+                * errorCode = 3U;
+            }
+            else 
+            {
+                readBytes = m_payloadLen;
+                memcpy(buffer, m_inputBuf, readBytes);
+
+                m_inputState = InputState::INIT;
+                m_received   = 0U;
+                m_payloadLen = 0U;
+            }
+        }
+    }
+
+    return readBytes;
+}
+
+size_t CustomRosTransportTcp::readInternal(uint8_t* buffer, size_t size, int timeout, uint8_t* errorCode)
+{
+    size_t remaining = size;
+
+    SimpleTimer readTimer;
+    bool        timeOver = false;
+
+    *errorCode = 0U;
+
+    readTimer.start(timeout);
+
+    while ((0 < remaining) && (false == timeOver))
+    {
+        int count = m_tcpClient.read(buffer, remaining);
+        if (-1 == count)
+        {
+            *errorCode = 1U;   /* Set error flag    */
+            remaining  = size; /* Return 0 on error.*/
+            break;
+        }
+
+        remaining -= count;
+        buffer += count;
+
+#if defined(TARGET_NATIVE)
+        /*
+         * SimpleTimer requires simulation time to progress
+         * otherwise this loop will block until a packet was received.
+         * TODO This is a workaround which needs to be discussed further.
+         */
+        Board::getInstance().stepTime();
+#endif
+        if (readTimer.isTimeout())
+        {
+            timeOver = true;
+        }
+    }
+
+    return size - remaining;
+}
+
+bool CustomRosTransportTcp::readSizePrefix(int timeout, uint8_t* errorCode)
+{
+    bool loop = true;
+
+    uint8_t prefix[2];
+
+    m_received = 0U;
+
+    switch (readInternal(prefix, sizeof(prefix), timeout, errorCode))
+    {
+    case 0: /* nothing received */
```

Comment:
Decorator in all cases missing. This comment is valid for all of the other readXY methods too. :-)

Reply:
Decorators added